### ISSUE TITLE
Implement List.partition alternative that accepts number of partitions rather than their size

### ISSUE THREAD

**Issue by gissuebot** [google/guava#451]
_[Original issue](https://code.google.com/p/guava-libraries/issues/detail?id=451) created by **Hannes.Schmidt** on 2010-10-16 at 10:44 PM_

---

The current List.partition() guarantees that all but the last partition are of the specified size and that the last one may be smaller but not empty. It would be nice to have a version of List.partition that guarantees that it returns a specified number of roughly equally sized partitions (their sizes may  differ by 1 but no more).

The attached class does exactly that. It's lacking proper argument checking. I also attached a minimal  test case to illustrate the invariant.

**Comment by kevinb9n** [comment]
We had quite the epic discussion about this 3 years ago.

The upshot of it all is that yes, there is currently no way to accomplish a partition into a specified number of partitions (rather than partition size). Conceptually, there should be; in real life it's curious that this complaint has not arisen more often.

However, imho we should still do the aforementioned research into how many current partition() users are attempting this using bad math.

**Comment by kevinb9n** [comment]
Looking over 100+ randomly-chosen usages of this method, I discovered only two cases where the user may have been trying to use math to get a desired number of partitions. So that concern is not going to be a good source of motivation for this.

I'm afraid I just don't see this as commonly needed enough to go in Guava.

**Comment by ddlatham** [comment]
Searching for this functionality in Guava and found this issue.  My use case is taking a key space over a dataset that was previously partitioned into n ordered partitions and wanting to repartition (redistribute?) it into a smaller number of partitions k as evenly as possible.  Very sad to see it was closed and just 15 days ago.

**Comment by sekyo** [comment]
Searching for this functionality in Guava too and finally implemented it myself in our project. Our needs were quite specific but for some reasons we need to divide a list of objects into a precise number of sublists (having the sizes of the sub lists as close as possible) in order to insert them in parallel into a database.

**Comment by kevinb9n** [comment]
The need for this keeps coming up again. I'm going to reopen. The hard part is figuring out the API since there are now a 2x2 matrix of partition styles. The Iterables class is also slowly dying.

One possibility is that this functionality could fall out naturally from designing a Grid<T> type, which is a dense Table<Integer, Integer, T>; I think *some* proportion of current Table use cases might want to use that, and it could also potentially provide the API for partitioning various ways.

**Comment by kevinb9n** [comment]
We should also research how many current partition users in Google's codebase might be buggy for lack of this functionality.

**Comment by paulferaud** [comment]
> We should also research how many current partition users in Google's codebase might be buggy for lack of this functionality.

FWIW, I have the somewhat common case of “I have n elements, and my API accepts k elements at most per call”.

I don’t have a “fixed bucket count” issue as much as “Split this up according to my max bucket size, but please distribute evenly among the buckets”.

So “balanced split” is what I usually go for. My code is usually just:

```java
    int count = myList.size();
    double chunks = Math.ceil(1.0 * count / maxSize);
    int chunkSize = (int) Math.ceil(count / chunks);
    List<List<Element>> elementsPartitioned = Lists.partition(myList, chunkSize);
```

This has the advantage of being “not a lot of code”. But “not a lot of code” is still more code than “a single function call”, and I’m always stressed that my math might be wrong. It’s not the kind of thing I’m thrilled about doing on a daily basis.

I’d be very happy with a `Lists.distribute`, even if the output was a *copy* of the input (So `List<List<> Collections.distribute(Iterable<>, int)` would also work for me).

Other thoughts:
1. My algo isn’t “homogeneously balanced”: The last list can have up to `chunks - 1` elements less than every other list. I honestly don’t care.
2. `Hashing.consistentHash` would also be another alternative. But in my use-case, it is *imperative* that no bucket ever has a size bigger than `maxSize`.
3. I assume that if you have N workers, you could turn this algorithm the other way around, with the caveat that you last worker could have up to N-1 tasks.
